server/planning/events/events.py

```diff
@@ -797,9 +798,73 @@ def _convert_to_recurring_event(self, updates, original):
         app.on_inserted_events(generated_events)
         return generated_events
 
-    def get_recurring_timeline(self, selected, spiked=False):
-        events_base_service = EventsBaseService("events", backend=superdesk.get_backend())
-        return events_base_service.get_recurring_timeline(selected, postponed=True, spiked=spiked)
+    def get_recurring_timeline(self, selected, spiked=False, postponed=True):
+        excluded_states = []
+
+        if not spiked:
+            excluded_states.append(WORKFLOW_STATE.SPIKED)
+        if not postponed:
+            excluded_states.append(WORKFLOW_STATE.POSTPONED)
+
+        query = {
+            "$and": [
+                {"recurrence_id": selected["recurrence_id"]},
+                {"_id": {"$ne": selected[ID_FIELD]}},
+            ]
+        }
+
+        if excluded_states:
+            query["$and"].append({"state": {"$nin": excluded_states}})
+
+        sort = '[("dates.start", 1)]'
+        max_results = get_max_recurrent_events()
+        selected_start = selected.get("dates", {}).get("start", utcnow())
+
+        # Make sure we are working with a datetime instance
+        if not isinstance(selected_start, datetime):
+            selected_start = datetime.strptime(selected_start, "%Y-%m-%dT%H:%M:%S%z")
+
+        historic = []
+        past = []
+        future = []
+
+        for event in self.get_series(query, sort, max_results):
+            event["dates"]["end"] = event["dates"]["end"]
+            event["dates"]["start"] = event["dates"]["start"]
+            for sched in event.get("_planning_schedule", []):
+                sched["scheduled"] = sched["scheduled"]
+            end = event["dates"]["end"]
+            start = event["dates"]["start"]
+            if end < utcnow():
+                historic.append(event)
+            elif start < selected_start:
+                past.append(event)
+            elif start > selected_start:
+                future.append(event)
+
+        return historic, past, future
+
+    def get_series(self, query, sort, max_results):
+        page = 1
+
+        while True:
+            # Get the results from mongo
+            req = ParsedRequest()
+            req.sort = sort
+            req.where = json.dumps(query)
+            req.max_results = max_results
+            req.page = page
+            results = self.get_from_mongo(req=req, lookup=None)
```

Comment:
This method can be converted to async by using `get_from_mongo_async`. See the [conversion chart](https://superdesk--2861.org.readthedocs.build/en/2861/tutorials/upgrade_eve_resource_to_async_service.html#upgrading-crud-lifecycle-service-methods) for what functions have an async version available